Evals that die under the debugger may or may not leak eval data depending on how they die.

**Description**
If eval string is used when the debugger is enabled then we may or may not leak debugging meta data in the stash depending on how the eval fails. 

When `$^P` is set to an integer with the `0x400` bit set (and maybe other bits) the debugger mode is enabled and when code is compiled a copy of the lines that were compiled is stashed away in `%::` in a slot named `"_<(eval $eval_ctr)"`, we make attempts to arrange that this slot is deleted if the eval fails. However there are a variety of ways that the compilation phase of eval may fail. If the failure triggers a `Perl_croak()` then we leak the debugging data and do not delete the entry.

Two easy ways to see this are to `eval "BEGIN{die}"`, and another is trigger 10 or more syntax errors in the eval text, which will cause the parser to trigger a croak. Any error pathway that triggers a croak may cause this outcome and there are a handful of special cases where the parser /will/ croak.

Thus `"1+"` will fail compile in a way that cleans up `"_<(eval $eval_ctr)"`, but `"1+;1+;1+;1+;1+;1+;1+;1+;1+;1+;"` will trigger a croak and the `"_<(eval $eval_ctr)"` will be left behind.

`t/comp/retainedlines.t` attempts to test this behavior, but it only uses very simple code to eval, and it doesnt trigger a croak, which exposes the issue. I noticed this because I am trying to make the parser stop after the first syntax error it encounters, and this is easiest done by croaking, but that exposes the underlying bug in this ticket which exists with or without the new logic.

**Steps to Reproduce**
This should output "0 0", it outputs "0 1" instead.

```
perl -le'for (1,10) { $^P=0x400; eval "1+;" x $_; my $count=0; /eval/ and $count++ for keys %::; push @ret,$count } print "@ret"'
0 1
```

**Expected behavior**
We should clean up the entries consistently. The output from the above script should be `"0 0"`.

Or I guess alternatively we shouldn't care about these "stash droppings". Although that seems like a great way to make the debugger leak memory. On the other hand, its not clear to me if we clean up these droppings if the eval is successful, so I am not sure if we should care. It feels like we should, but maybe not.

**Perl configuration**
```
$ ./perl -Ilib -V
Summary of my perl5 (revision 5 version 37 subversion 4) configuration:
  Derived from: ebc2efe6583f7cf7ea29a7f94f7b7887a75ad3a7
  Ancestor: 18fa8a6f818cbe2838cfe9b1bfa0c5d9c311930c
  Platform:
    osname=linux
    osvers=5.14.0-1049-oem
    archname=x86_64-linux-thread-multi
    uname='linux oncidium 5.14.0-1049-oem #56-ubuntu smp fri aug 12 10:23:08 utc 2022 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dusethreads -Doptimize=-g -d -Dusedevel -Dcc=ccache gcc -Dld=gcc -DDEBUGGING'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-g'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='9.4.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='gcc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib /usr/lib64
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.31.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.31'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -g -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    DEBUGGING
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_TRACK_MEMPOOL
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Locally applied patches:
    uncommitted-changes
    b3a0a5e1356f495dc935458a9ee5b38085e130f8
    f1886e10a63297209ad9c385b054717c4a089d8b
    ebc2efe6583f7cf7ea29a7f94f7b7887a75ad3a7
  Built under linux
  Compiled at Aug 26 2022 14:32:36
  %ENV:
    PERLBREW_CONFIGURE_FLAGS="-de -Dcc=ccache\ gcc -Dld=gcc"
    PERLBREW_HOME="/home/yorton/.perlbrew"
    PERLBREW_MANPATH="/home/yorton/perl5/perlbrew/perls/perl-5.34.1/man"
    PERLBREW_PATH="/home/yorton/perl5/perlbrew/bin:/home/yorton/perl5/perlbrew/perls/perl-5.34.1/bin"
    PERLBREW_PERL="perl-5.34.1"
    PERLBREW_ROOT="/home/yorton/perl5/perlbrew"
    PERLBREW_SHELLRC_VERSION="0.88"
    PERLBREW_VERSION="0.88"
  @INC:
    lib
    /usr/local/lib/perl5/site_perl/5.37.4/x86_64-linux-thread-multi
    /usr/local/lib/perl5/site_perl/5.37.4
    /usr/local/lib/perl5/5.37.4/x86_64-linux-thread-multi
    /usr/local/lib/perl5/5.37.4
```

I think a few folks have been doing research, so there may be some duplication of effort here. But I have focusing on the fact that CATCH_GET seems to never be true when called from pp_entereval or pp_require. Both have docatch() wrappers which clear the flag. However if I change the clause in doeval_compile():

```
-    yystatus = (!in_require && CATCH_GET) ? S_try_yyparse(aTHX_ GRAMPROG) : yyparse(GRAMPROG);
+    yystatus = (!in_require || CATCH_GET)   ? S_try_yyparse(aTHX_ GRAMPROG) : yyparse(GRAMPROG);
```

Or if i remove the CATCH_GET entirely I end up with the following test fails:

```
Test Summary Report
-------------------
../cpan/IO-Compress/t/020isize.t                                   (Wstat: 139 (Signal: SEGV, dumped core) Tests: 0 Failed: 0)
  Non-zero wait status: 139
../cpan/IO-Compress/t/105oneshot-zip-store-only.t                  (Wstat: 139 (Signal: SEGV, dumped core) Tests: 0 Failed: 0)
  Non-zero wait status: 139
../cpan/autodie/t/import-into.t                                    (Wstat: 512 (exited 2) Tests: 0 Failed: 0)
  Non-zero exit status: 2
../cpan/version/t/11_taint.t                                       (Wstat: 512 (exited 2) Tests: 0 Failed: 0)
  Non-zero exit status: 2
../dist/Storable/t/leaks.t                                         (Wstat: 512 (exited 2) Tests: 0 Failed: 0)
  Non-zero exit status: 2
Files=2720, Tests=1187179, 336 wallclock secs (227.51 usr 14.93 sys + 1840.35 cusr 72.46 csys = 2155.25 CPU)
Result: FAIL
make: *** [makefile:856: test_harness] Error 6
```

Both however do fix the problem with t/comp/retained_lines.pl

In my previous message I meant to say the "first JMPENV_PUSH()".

Also now that i can see that the pieces are starting to come into place.

This is from https://github.com/Perl/perl5/blob/blead/perl.c#L1726

```
For historical reasons, the non-zero return value also attempts to
be a suitable value to pass to the C library function C<exit> (or to
return from C<main>), to serve as an exit code indicating the nature
of the way initialisation terminated.  However, this isn't portable,
due to differing exit code conventions.  A historical bug is preserved
for the time being: if the Perl built-in C<exit> is called during this
function's execution, with a type of exit entailing a zero exit code
under the host operating system's conventions, then this function
returns zero rather than a non-zero value.  This bug, [perl #2754],
leads to C<perl_run> being called (and therefore C<INIT> blocks and the
main program running) despite a call to C<exit>.  It has been preserved
because a popular module-installing module has come to rely on it and
needs time to be fixed.  This issue is [perl #132577], and the original
bug is due to be fixed in Perl 5.30.
```

and if we look at this:

```
$ ./perl -Dl -wle 'BEGIN { eval qq( BEGIN { warn "A"; die "A"; }); print "ERR: $@"; exit; }' 2>&1 | tail -n 20
(-e:1)	LEAVE scope 2 (savestack=0) at perl.c:1919
(-e:1)	JMPENV_POP level=1 at perl.c:1945
(-e:1)	JMPENV_PUSH level=1 ret=0 at perl.c:2696
```
This here ^^^
```
(-e:1)	Entering new RUNOPS level
(-e:1)	ENTER scope 2 (savestack=0) at dump.c:2703
(-e:1)	LEAVE scope 2 (savestack=2) at dump.c:2723
(-e:1)	savestack: releasing items 2 -> 0
panic: freed op 0x564d359fbaf0 called
(-e:1)	JMPENV_JUMP(2) level=1 at perl.c:5392
(-e:1)	JMPENV_PUSH level=1 ret=2 at perl.c:2696
(-e:1)	JMPENV_POP level=1 at perl.c:2732
(-e:1)	JMPENV_PUSH level=1 ret=0 at perl.c:663
(-e:1)	JMPENV_POP level=1 at perl.c:669
(-e:1)	LEAVE scope 1 (savestack=0) at perl.c:671
ERR: A at (eval 1) line 1.
BEGIN failed--compilation aborted at (eval 1) line 1.

(-e:1)	ENTER scope 1 (savestack=0) at pad.c:344
(-e:1)	LEAVE scope 1 (savestack=2) at pad.c:353
(-e:1)	savestack: releasing items 2 -> 0
```

That comes from this:

```
    if (!perl_parse(my_perl, xs_init, argc, argv, (char **)NULL))
        perl_run(my_perl);
```

so what i think is happening is that we call exit(0);  this then falls into the commented logic, and we run the run loop anyway, and then that segfaults because the opcodes don't exist anymore. This seems borne out by the fact that changing the code to do an exit(1) instead of 0 eliminates the error.

--- a/cop.h
+++ b/cop.h
@@ -116,8 +116,8 @@ typedef struct jmpenv JMPENV;
             int i = 0;                                                  \
             JMPENV *p = PL_top_env;                                     \
             while (p) { i++; p = p->je_prev; }				\
-            Perl_deb(aTHX_ "JMPENV_PUSH pre level=%d at %s:%d\n",       \
-                         i,  __FILE__, __LINE__);                       \
+            Perl_deb(aTHX_ "JMPENV_PUSH pre level=%d in %s at %s:%d\n", \
+                         i,  SAFE_FUNCTION__, __FILE__, __LINE__);      \
         });                                                             \
         cur_env.je_prev = PL_top_env;					\
         JE_OLD_STACK_HWM_save(cur_env);                                 \
@@ -130,8 +130,8 @@ typedef struct jmpenv JMPENV;
             int i = 0;                                                  \
             JMPENV *p = PL_top_env;                                     \
             while (p) { i++; p = p->je_prev; }				\
-            Perl_deb(aTHX_ "JMPENV_PUSH level=%d ret=%d at %s:%d\n",    \
-                         i, cur_env.je_ret,  __FILE__, __LINE__);       \
+            Perl_deb(aTHX_ "JMPENV_PUSH level=%d ret=%d in %s at %s:%d\n",    \
+                         i, cur_env.je_ret, SAFE_FUNCTION__,  __FILE__, __LINE__); \
         });                                                             \
         (v) = cur_env.je_ret;						\
     } STMT_END
@@ -141,8 +141,8 @@ typedef struct jmpenv JMPENV;
         DEBUG_l({                                                       \
             int i = -1; JMPENV *p = PL_top_env;				\
             while (p) { i++; p = p->je_prev; }				\
-            Perl_deb(aTHX_ "JMPENV_POP level=%d at %s:%d\n",		\
-                         i, __FILE__, __LINE__);})			\
+            Perl_deb(aTHX_ "JMPENV_POP level=%d in %s at %s:%d\n",        \
+                         i, SAFE_FUNCTION__, __FILE__, __LINE__);})        \
         assert(PL_top_env == &cur_env);					\
         PL_delaymagic = cur_env.je_old_delaymagic;			\
         PL_top_env = cur_env.je_prev;					\
@@ -153,8 +153,8 @@ typedef struct jmpenv JMPENV;
         DEBUG_l({                                               \
             int i = -1; JMPENV *p = PL_top_env;			\
             while (p) { i++; p = p->je_prev; }			\
-            Perl_deb(aTHX_ "JMPENV_JUMP(%d) level=%d at %s:%d\n", \
-                         (int)(v), i, __FILE__, __LINE__);})    \
+            Perl_deb(aTHX_ "JMPENV_JUMP(%d) level=%d in %s at %s:%d\n",         \
+                         (int)(v), i, SAFE_FUNCTION__, __FILE__, __LINE__);})   \
         if (PL_top_env->je_prev)				\
             PerlProc_longjmp(PL_top_env->je_buf, (v));		\
         if ((v) == 2)						\
@@ -168,9 +168,9 @@ typedef struct jmpenv JMPENV;
     STMT_START {							\
         DEBUG_l(                                                        \
             Perl_deb(aTHX_						\
-                "JUMPLEVEL set catch %d => %d (for %p) at %s:%d\n",	\
+                "JUMPLEVEL set catch %d => %d (for %p) in %s at %s:%d\n",   \
                  PL_top_env->je_mustcatch, (v), (void*)PL_top_env,      \
-                 __FILE__, __LINE__);)					\
+                 SAFE_FUNCTION__, __FILE__, __LINE__);)			\
         PL_top_env->je_mustcatch = (v);					\
     } STMT_END
 
@@ -889,7 +889,7 @@ struct block {
 
 #define CX_DEBUG(cx, action)						\
     DEBUG_l(								\
-        Perl_deb(aTHX_ "CX %ld %s %s (scope %ld,%ld) (save %ld,%ld) at %s:%d\n",\
+        Perl_deb(aTHX_ "CX %ld %s %s (scope %ld,%ld) (save %ld,%ld) in %s at %s:%d\n",\
                     (long)cxstack_ix,					\
                     action,						\
                     PL_block_type[CxTYPE(cx)],	                        \
@@ -897,7 +897,7 @@ struct block {
                     (long)(cx->blk_oldscopesp),		                \
                     (long)PL_savestack_ix,				\
                     (long)(cx->blk_oldsaveix),                          \
-                    __FILE__, __LINE__));
+                    SAFE_FUNCTION__, __FILE__, __LINE__));
 
 
 
@@ -1162,8 +1162,8 @@ typedef struct stackinfo PERL_SI;
         DEBUG_l({							\
             int i = 0; PERL_SI *p = PL_curstackinfo;			\
             while (p) { i++; p = p->si_prev; }				\
-            Perl_deb(aTHX_ "push STACKINFO %d at %s:%d\n",		\
-                         i, __FILE__, __LINE__);})			\
+            Perl_deb(aTHX_ "push STACKINFO %d in %s at %s:%d\n",        \
+                         i, SAFE_FUNCTION__, __FILE__, __LINE__);})        \
         if (!next) {							\
             next = new_stackinfo(32, 2048/sizeof(PERL_CONTEXT) - 1);	\
             next->si_prev = PL_curstackinfo;				\
@@ -1190,8 +1190,8 @@ typedef struct stackinfo PERL_SI;
         DEBUG_l({							\
             int i = -1; PERL_SI *p = PL_curstackinfo;			\
             while (p) { i++; p = p->si_prev; }				\
-            Perl_deb(aTHX_ "pop  STACKINFO %d at %s:%d\n",		\
-                         i, __FILE__, __LINE__);})			\
+            Perl_deb(aTHX_ "pop  STACKINFO %d in %s at %s:%d\n",        \
+                         i, SAFE_FUNCTION__, __FILE__, __LINE__);})        \
         if (!prev) {							\
             Perl_croak_popstack();					\
         }								\

--- a/cop.h
+++ b/cop.h
@@ -112,10 +112,12 @@ typedef struct jmpenv JMPENV;
 
 #define JMPENV_PUSH(v)                                                  \
     STMT_START {							\
-        PERL_DEB(int i=0);                                              \
         DEBUG_l({                                                       \
+            int i = 0;                                                  \
             JMPENV *p = PL_top_env;                                     \
             while (p) { i++; p = p->je_prev; }				\
+            Perl_deb(aTHX_ "JMPENV_PUSH pre level=%d at %s:%d\n",       \
+                         i,  __FILE__, __LINE__);                       \
         });                                                             \
         cur_env.je_prev = PL_top_env;					\
         JE_OLD_STACK_HWM_save(cur_env);                                 \
@@ -124,11 +126,14 @@ typedef struct jmpenv JMPENV;
         PL_top_env = &cur_env;						\
         cur_env.je_mustcatch = FALSE;					\
         cur_env.je_old_delaymagic = PL_delaymagic;			\
-        (v) = cur_env.je_ret;						\
         DEBUG_l({                                                       \
+            int i = 0;                                                  \
+            JMPENV *p = PL_top_env;                                     \
+            while (p) { i++; p = p->je_prev; }				\
             Perl_deb(aTHX_ "JMPENV_PUSH level=%d ret=%d at %s:%d\n",    \
                          i, cur_env.je_ret,  __FILE__, __LINE__);       \
         });                                                             \
+        (v) = cur_env.je_ret;						\
     } STMT_END
 
 #define JMPENV_POP \

--- a/handy.h
+++ b/handy.h
@@ -132,10 +132,13 @@ required, but is kept for backwards compatibility.
  * XXX Similarly, a Configure probe for __FILE__ and __LINE__ is needed. */
 #if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L) || (defined(__SUNPRO_C)) /* C99 or close enough. */
 #  define FUNCTION__ __func__
+#  define SAFE_FUNCTION__ __func__
 #elif (defined(__DECC_VER)) /* Tru64 or VMS, and strict C89 being used, but not modern enough cc (in Tur64, -c99 not known, only -std1). */
-#  define FUNCTION__ ""
+#  define FUNCTION__ ("")
+#  define SAFE_FUNCTION__ ("UNKNOWN")
 #else
 #  define FUNCTION__ __FUNCTION__ /* Common extension. */
+#  define SAFE_FUNCTION__ __FUNCTION__ /* Common extension. */
 #endif
 
 /* XXX A note on the perl source internal type system.  The

--- a/pp_ctl.c
+++ b/pp_ctl.c
@@ -3752,7 +3752,55 @@ S_doeval_compile(pTHX_ U8 gimme, CV* outside, U32 seq, HV *hh)
 
     if (PL_unitcheckav) {
         OP *es = PL_eval_start;
-        call_list(PL_scopestack_ix, PL_unitcheckav);
+        if (in_require) {
+            call_list(PL_scopestack_ix, PL_unitcheckav);
+        } else {
+            /* TODO: are we sure we shouldn't do JMPENV_PUSH in
+             * when `in_require` is true? */
+            int ret=0;
+            dJMPENV;
+            JMPENV_PUSH(ret);
+            switch (ret) {
+            case 0:
+                /*
+                 * Doesn't seem like PUSHMARK(SP)/ENTER
+                 * is needed here. */
+
+                call_list(PL_scopestack_ix, PL_unitcheckav);
+                /* Nor LEAVE here. */
+                break;
+            case 3: {
+                /* call_list failed and threw an error */
+
+                /* Restore PL_OP */
+                PL_op = saveop;
+
+                SV *errsv = ERRSV;
+                if (!*(SvPV_nolen_const(errsv))) {
+                    /* This happens when using:
+                     * eval qq# UNITCHECK { die "\x00"; } #;
+                     */
+                    sv_setpvs(errsv, "Unit check error");
+                }
+
+                /* We're returning so POP our JMPENV */
+                /* NOTE: in `S_try_yyparse` the default for ret=3 is to
+                 *       break which falls back to the `JMPENV_POP`
+                 *       after the switch. In this code we're returning
+                 *       early so we must POP it outrself. */
+                JMPENV_POP;
+
+                if (gimme != G_LIST) PUSHs(&PL_sv_undef);
+                PUTBACK;
+                return FALSE;
+              }
+            default:
+                JMPENV_POP;
+                JMPENV_JUMP(ret);
+                NOT_REACHED; /* NOTREACHED */
+            }
+            JMPENV_POP;
+        }
         PL_eval_start = es;
     }
 

--- a/pp_ctl.c
+++ b/pp_ctl.c
@@ -3686,12 +3686,13 @@ S_doeval_compile(pTHX_ U8 gimme, CV* outside, U32 seq, HV *hh)
 
     CALL_BLOCK_HOOKS(bhk_eval, saveop);
 
-    /* note that yyparse() may raise an exception, e.g. C<BEGIN{die}>,
-     * so honour CATCH_GET and trap it here if necessary */
-
-
+    /* we should never be CATCH_GET true here, as our immediate callers should
+     * always handle that case. */
+    assert(!CATCH_GET);
     /* compile the code */
-    yystatus = (!in_require && CATCH_GET) ? S_try_yyparse(aTHX_ GRAMPROG) : yyparse(GRAMPROG);
+    yystatus = (!in_require)
+               ? S_try_yyparse(aTHX_ GRAMPROG)
+               : yyparse(GRAMPROG);
 
     if (yystatus || PL_parser->error_count || !PL_eval_root) {
         PERL_CONTEXT *cx;
@@ -4577,6 +4578,8 @@ PP(pp_entereval)
      */
     if (CATCH_GET)
         return docatch(Perl_pp_entereval);
+
+    assert(!CATCH_GET);
 
     gimme = GIMME_V;
     was = PL_breakable_sub_gen;

--- a/pp_ctl.c
+++ b/pp_ctl.c
@@ -3750,7 +3750,7 @@ S_doeval_compile(pTHX_ U8 gimme, CV* outside, U32 seq, HV *hh)
         }
     }
 
-    if (PL_unitcheckav) {
+    if (PL_unitcheckav && av_count(PL_unitcheckav)>0) {
         OP *es = PL_eval_start;
         if (in_require) {
             call_list(PL_scopestack_ix, PL_unitcheckav);

--- a/t/comp/retainedlines.t
+++ b/t/comp/retainedlines.t
@@ -6,7 +6,7 @@
 # we've not yet verified that use works.
 # use strict;
 
-print "1..98\n";
+print "1..109\n";
 my $test = 0;
 
 sub failed {
@@ -164,6 +164,26 @@ for (0xA, 0) {
        "evals with BEGIN{die} are correctly cleaned up");
   }
 }
+
+for (0xA, 0) {
+  local $^P = $_;
+
+  eval (my $prog = "UNITCHECK{die}\n");
+  is (!!$@, 1, "Is \$@ true?");
+  is ($@=~/UNITCHECK failed--call queue aborted/, 1,
+      "Error is expected value?");
+
+  if ($_) {
+    check_retained_lines($prog, 'eval that defines UNITCHECK that dies');
+  }
+  else {
+    my @after = grep { /eval/ } keys %::;
+
+    is (scalar @after, 0 + keys %seen,
+       "evals with UNITCHECK{die} are correctly cleaned up");
+  }
+}
+
 
 # [perl #79442] A #line "foo" directive in a string eval was not updating
 # *{"_<foo"} in threaded perls, and was not putting the right lines into

--- a/t/comp/retainedlines.t
+++ b/t/comp/retainedlines.t
@@ -6,7 +6,7 @@
 # we've not yet verified that use works.
 # use strict;
 
-print "1..75\n";
+print "1..98\n";
 my $test = 0;
 
 sub failed {
@@ -101,7 +101,10 @@ for my $sep (' ', "\0") {
   my $prog = "sub $name {
     'This is $name'
   }
-1 +
+# 10 errors to triger a croak during compilation.
+1 +; 1 +; 1 +; 1 +; 1 +;
+1 +; 1 +; 1 +; 1 +; 1 +;
+1 +; # and one more for good measure.
 ";
 
   eval $prog and die;
@@ -119,7 +122,9 @@ foreach my $flags (0x0, 0x800, 0x1000, 0x1800) {
     # This is easier if we accept that the guts eval will add a trailing \n
     # for us
     my $prog = "1 + 1 + 1\n";
-    my $fail = "1 + \n";
+    my $fail = "1 +;\n" x 11; # we need 10 errors to trigger a croak during
+                              # compile, we add an extra one just for good
+                              # measure.
 
     is (eval $prog, 3, 'String eval works');
     if ($flags & 0x800) {

--- a/t/op/eval.t
+++ b/t/op/eval.t
@@ -6,7 +6,7 @@ BEGIN {
     set_up_inc('../lib');
 }
 
-plan(tests => 146);
+plan(tests => 149);
 
 eval 'pass();';
 
@@ -715,4 +715,15 @@ pass("eval in freed package does not crash");
         'Blocked BEGIN results in expected error');
     ::is($x,2,'BEGIN really did nothing');
 
+}
+
+{
+    # make sure that none of these segfault.
+    foreach my $line (
+        'eval "UNITCHECK { eval q(UNITCHECK { die; }); print q(A-) }";',
+        'eval "UNITCHECK { eval q(BEGIN     { die; }); print q(A-) }";',
+        'eval "BEGIN     { eval q(UNITCHECK { die; }); print q(A-) }";',
+    ) {
+        fresh_perl_is($line . ' print "ok";', "A-ok", {}, "No segfault: $line");
+    }
 }

